### PR TITLE
feat: show course module icons in course index, resolves #16

### LINKED ISSUE #16: Feature: Display activity icons in course content navigation
In order to provide quick visual orientation for users it would be helpful to integrate activity icons in the course content menu on the left hand side. Right now in Boost 4.0 this place has been reserved for the (additional) indication of the activity completion status. In order to quickly find the activity I am looking for, I am obliged to read all the titles, which are manually added by teachers and will never be consistent throughout courses.

Speaking on behalf of our organisation (Hochschule Hannover, Higher education) very few teachers actually use the activity completion, but everybody navigates through course content. 

Discussion is needed on how to best solve this:

Maybe this features could at least be an optional configuration for teachers (show activity completion in course navigation, show activity icons).

activity-icons

### PR DESCRIPTION
**Pull request by christianwolters**

### CODE REVIEW

File: lang/en/theme_boost_union.php [hunk 1]
```diff
@@ -242,6 +242,15 @@
 $string['courseheaderimagelayoutheadingabove'] = 'Course title above of course image';
 $string['courseheaderimageposition'] = 'Course header image position';
 $string['courseheaderimageposition_desc'] = 'With this setting, you control the positioning of the course header image within the course header image container. The first value is the horizontal position, the second value is the vertical position.';
+// ... Section: Course index.
+$string['courseindexheading'] = 'Course Index';
+// ... ... Setting: Course index.
+$string['courseindexmodiconenabled'] = 'Display course module icons in course index';
+$string['courseindexmodiconenabled_desc'] = 'When enabled, the corresponding course module icon is displayed in front of the index row. In doing so, the course module icon is replacing the course completion indicator which is moved from the front to the right side of the index row.';
```
Comment (lucaboesch) [inline lang/en/theme_boost_union.php:249]: `$string['courseindexmodiconenabled_desc'] = 'When enabled, the corresponding course module icon is displayed in front of the index row. In doing so, the course module icon is replacing the course module completion indicator which is moved from the front to the right side of the index row.'`

Better not to confuse course completion indicator and course module completion indicator.

**Comment by lucaboesch** [comment]
Thanks for your proposal, @christianwolters !

Allow me some thoughts about your work.
You offer two options which are carefully programmed for. From default to the left hand side to "End of line" (middle) and "Icon" (right)
image

Now, for me there are options left out, and I don't know whether because it is impossible to accomplish or why else but I'd like to mix them into the game. That would be (default to the left hand side again) "Start of line" which also could be called "in front of the icon" (middle) and "after the icon" (right)

image

So, in the end, the four options could be: "in front of the icon", "after the icon", "end of line", and "colour in the icon".

What do you think?

Best,
Luca

**Comment by christianwolters** [comment]
Hi @lucaboesch,

thanks for taking the time to look at this.

Regarding the "left out" options: 
* my approach was to deliver a first step into course-index manipulation in boost union
* the options provided for the completion indication were the first that came to my mind
* personally, displaying icon and completion circle side-by-side is a bit crowded (but that is really IMHO)

That being said, I like the idea of offering more display options (and also i really like your option naming: "in front of the icon", "after the icon", "end of line", and "colour in the icon"). 

Right now, I'm thinking that these options would me implementable, but there is certainly a need for some thinking and refactoring, which would block this PR. 

So I'd like to move your ideas into a new issue, so they become the next iteration of this feature?

Cheers 
Christian

File: templates/core_courseformat/local/courseindex/cmcompletion.mustache [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        "state": 1,
+        "iscomplete": true,
+        "isincomplete": true,
+        "isfail": true,
+        "hasstate": true
+    }
+}}
+{{!
+    Modifications of upstream
+    * TODO
```
Comment (christianwolters) [inline templates/core_courseformat/local/courseindex/cmcompletion.mustache:33]: Ooops...

File: lang/en/theme_boost_union.php (hunk 1 above)
Comment (lucaboesch) [reply lang/en/theme_boost_union.php:249]: It seem to me that watching https://www.youtube.com/watch?v=PHvklKLq2T4 I think it's best to talk only of 'activities'.
The double mentioning of 'activities and resources' is to be dropped and 'resources' is just being seen as a 'category' of an activity, such as 'assessment', 'collaboration', 'communication' et cetera.

So, me, personally would just call it 'activity' what you see as single course module, e.g. 'quiz 1', 'assessment abc', 'workshop α' and 'activity type' when referencing to what influences the icon, e.g. quiz, assessment, workshop.

In that spirit:
`
$string['courseindexacttypeiconenabled'] = 'Display activity type icons in course index';
$string['courseindexacttypeiconenabled_desc'] = 'When enabled, the corresponding activity type icon is displayed in front of the index row. In doing so, the course activity type is replacing the course completion indicator which is moved from the front to the right side of the index row.';
`

**Comment by lucaboesch** [comment]
Yes, sure, it is very welcome to add the two other options as follow-up issues.
It is unlikely they would have been in in time for our upgrade February 14th, 2024, anyway 😜.

**Comment by abias** [comment]
Hi @christianwolters ,
many thanks for contributing this feature!

And thanks @lucaboesch for your initial thoughts!

I have now taken this over and made an integration review.

Doing this:
* I have raised version in version.php (as you have introduced new settings)
* I added a note to CHANGES.md
* I added a section to README.md
* I improved the overall coding style and verbosity
* I reverted some indention changes in the duplicated mustache templates to shrink the diff between the BU template and the upstream template
* I fixed the example context in cm.mustache
* I fixed the drag icon in edit mode (which you broke unfortunately)
* I improved the alignment of the activity icons in connection with the section titles
* I added an icon style for dimmed / inactive activities
* I added the option "in front of the icon" as Luca suggested it (I just had to ignore "after the icon" for now due to time constraints)
* I have extended the Behat tests and have transformed it into a Scenenario outline. It's not perfect, as this patch has massive visual impact which can't be tested with Behat, but it should be fine for the time being.

Unfortunately, I was not able to push my review changes to your fork @christianwolters.

That's why I will close this PR now and will create a follow-up PR on #562.

Cheers,
Alex

**Pull request closed without merging**